flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java

```diff
@@ -343,13 +412,66 @@ public Builder field(String name, TypeInformation<?> typeInfo) {
 			return field(name, fromLegacyInfoToDataType(typeInfo));
 		}
 
+		/**
+		 * Add a primary key with the given field names.
+		 * There can only be one PRIMARY KEY for a given table
+		 * See the {@link TableSchema} class javadoc for more definition about primary key.
+		 */
+		public Builder primaryKey(String... fields) {
+			Preconditions.checkArgument(
+				fields != null && fields.length > 0,
+				"The primary key fields shouldn't be null or empty.");
+			Preconditions.checkArgument(
+				primaryKey == null,
+				"A primary key " + primaryKey +
+					" have been defined, can not define another primary key " +
+					Arrays.toString(fields));
+			for (String field : fields) {
+				if (!fieldNames.contains(field)) {
+					throw new IllegalArgumentException("The field '" + field +
+						"' is not existed in the schema.");
+				}
+			}
```

Comment:
that means we must build fieldNames first ?

Reply:
Yes. I think so.